File: 누적합/실버/실버5_2167_2차원 배열의 합/matrix.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin', 'utf8').trim().split(/\s+/);
const lines = input.trim().split('\n');
let idx = 0;

// Read N and M
const [N, M] = lines[idx].split(' ').map(Number);
idx += 1;

// Read the array
const array = [];
for (let i = 0; i < N; i++) {
array.push(lines[idx].split(' ').map(Number));
idx += 1;
    }

    // Create the prefix sum array
function solution(array, N, M) {
    const prefixSum = Array.from(Array(N + 1), () => Array(M + 1).fill(0));

    // Fill the prefix sum array
    for (let i = 1; i <= N; i++) {
        for (let j = 1; j <= M; j++) {
            prefixSum[i][j] = array[i - 1][j - 1] +
                              prefixSum[i - 1][j] +
                              prefixSum[i][j - 1] -
                              prefixSum[i - 1][j - 1];
        }
    }

    // Read K
    const K = Number(lines[idx]);
    idx += 1;

    const results = [];
    for (let k = 0; k < K; k++) {
        const [i, j, x, y] = lines[idx].split(' ').map(Number);
        idx += 1;

        // Calculate the sum using the prefix sum array
        const sum = prefixSum[x][y] -
                    prefixSum[i - 1][y] -
                    prefixSum[x][j - 1] +
                    prefixSum[i - 1][j - 1];

        results.push(sum);
    }

    // Output all results
    console.log(results.join('\n'));
}

solution(array, N, M);

File: 누적합/실버/실버5_2167_2차원 배열의 합/matrix.py
import sys
input = sys.stdin.read()
data = input.split('\n')

# 입력값 받기
N, M= list(map(int, data[0].split()))  # row, column
A = []
for i in range(1, N + 1):
    row = list(map(int, data[i].split()))
    A.append(row)

K = int(data[N + 1])

# prefix_sum 초기화
prefix_sum = [0] * (N + 1)  # row
for i in range(0, N + 1):  # column
    prefix_sum[i] = [0] * (M + 1)

# prefix_sum 구축
for i in range(1, N + 1):
    for j in range(1, M + 1):
        prefix_sum[i][j] = A[i - 1][j - 1] + prefix_sum[i - 1][j] + prefix_sum[i][j - 1] - prefix_sum[i - 1][j - 1]

# queries 값 정의
queries = []
for i in range(N + 2, N+ 2+ K):
    query = list(map(int, (data[i].split())))
    queries.append(query)

# query별 결과값 도출
for query in queries:
    i, j, x, y = query
    result = prefix_sum[x][y] - prefix_sum[i - 1][y] - prefix_sum[x][j - 1] + prefix_sum[i - 1][j - 1]
    print(result)
